### PR TITLE
Is subtype disallow extra struct fields

### PR DESCRIPTION
**Pull request by rafael-atticlabs**
It bothered me too much to copy this code in Attic, so I added another kinda weird type function to noms.

Still need to add tests, but let me know if you are ok with this approach.

FYI, I refactored the struct section because the existing code was a bit too hard to reason about. I know my version is longer but it's easier to convince myself that it's correct.

### CODE REVIEW

**Comment by rafael-atticlabs** [comment]
Tests added

**Comment by arv** [comment]
Another way to factor this might be to pass in a callback that is called before (and another one for after?) issubtype is called at a level. Then it would be easy to add custom logic for attic.

File: go/types/subtype.go [hunk 1]
```diff
@@ -16,18 +16,23 @@ func assertSubtype(t *Type, v Value) {
 
 // IsSubtype determines whether concreteType is a subtype of requiredType. For example, `Number` is a subtype of `Number | String`.
 func IsSubtype(requiredType, concreteType *Type) bool {
-	return isSubtype(requiredType, concreteType, nil)
+	return isSubtype(requiredType, concreteType, true, nil)
 }
 
-func isSubtype(requiredType, concreteType *Type, parentStructTypes []*Type) bool {
+// IsSubtypeDisallowExtraFields is a slightly wiered variant of IsSubtype. It returns true IFF IsSubtype(requiredType, concreteType) AND Structs in concreteType CANNOT have field names absent in requiredType
```
Comment (arv) [inline go/types/subtype.go]: typo... wiered -> weird

File: go/types/subtype.go [hunk 2]
```diff
@@ -215,7 +236,7 @@ func IsValueSubtypeOf(v Value, t *Type) bool {
 func isMetaSequenceSubtypeOf(ms metaSequence, t *Type) bool {
 	for _, mt := range ms.tuples {
 		// Each prolly tree is also a List<T> where T needs to be a subtype.
-		if !isSubtype(t, mt.ref.TargetType(), nil) {
+		if !isSubtype(t, mt.ref.TargetType(), true, nil) {
```
Comment (arv) [inline go/types/subtype.go]: Maybe add a third isSubtype without the bool and nil?

File: go/types/subtype_test.go [hunk 3]
```diff
@@ -262,8 +262,8 @@ func TestAssertTypeCycleUnion(tt *testing.T) {
 		StructField{"y", MakeUnionType(NumberType, StringType), false},
 	)
 
-	assert.True(tt, isSubtype(t2, t1, nil))
-	assert.False(tt, isSubtype(t1, t2, nil))
+	assert.True(tt, isSubtype(t2, t1, true, nil))
```
Comment (arv) [inline go/types/subtype_test.go]: Yeah, I don't like the extra params. It is too hard to reason about if you are not aware of the motivation of this weird case.

File: go/types/subtype_test.go [hunk 4]
```diff
@@ -438,6 +420,63 @@ func TestIsSubtypeOptionalFields(tt *testing.T) {
 	assert.False(IsSubtype(t2, t1))
 }
 
+func makeType(s string) *Type {
```
Comment (arv) [inline go/types/subtype_test.go]: maybe rename now that all tests see this.

File: go/types/subtype.go (hunk 1 above)
Comment (rafael-atticlabs) [inline go/types/subtype.go]: done

File: go/types/subtype.go (hunk 2 above)
Comment (rafael-atticlabs) [inline go/types/subtype.go]: done

File: go/types/subtype_test.go (hunk 4 above)
Comment (rafael-atticlabs) [inline go/types/subtype_test.go]: done

**Comment by rafael-atticlabs** [comment]
All done. TODO'd suggestion of adding a delegation method.

**Pull request merged**